### PR TITLE
[SuperTextField][Android] Fix glitch with predictive text (Resolves #1113)

### LINKED ISSUE #1113: [SuperTextField][Android] - Predictive Text Strange Behaviors
Super Editor - stable branch
Samsung Galaxy S21 FE
Android 13
One UI 5.1

There are weird responses from super text field, when user's predictive text is enabled. When you play a few minutes, you would notice a couple of bugs. @matthew-carroll @angelosilvestre @miguelcmedeiros  

1. When I delete the word's last character and put the same character again. Curser is not at the end, it is 1 character left.
It is happening current word. If I write new word look 2:

<details>
  <summary>Cursor Position - Logs Fine</summary>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-6aad8d2f-7fff-a7b6-94bb-f379bcf88d0b"><div dir="ltr" style="margin-left:0pt;" align="left">

I  Initializing logger: textfield
I  (31.948) textfield.android > INFO: Attaching TextInputClient to TextInput
I  (31.949) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy├, selection: TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (31.950) textfield.ime > FINE: Is attached to input client? true
I  (33.686) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning gu├, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (34.944) textfield.ime > FINE: Received text editing deltas from platform...
I  (34.945) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#88c2a(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (34.945) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#03871(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 0, end: 15))
I  (34.948) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#d15b5(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (34.949) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#c86a4(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (34.950) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#faaef(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 15))
I  (34.953) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning gu├, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 15))
I  (34.954) textfield.ime > FINE: Received text editing deltas from platform...
I  (34.955) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#cc7fd(oldText: Good morning gu, textInserted: y, insertionOffset: 15, selection: TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 16))
I  (34.956) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy├, selection: TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 16))
I  (34.965) textfield.ime > FINE: Received text editing deltas from platform...
I  (34.966) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#b7ac1(oldText: Good morning gu, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 15))
I  (34.968) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy├, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 15))
I  (34.976) textfield.ime > FINE: Received text editing deltas from platform...
I  (34.977) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#22c4d(oldText: Good morning guy, selection: TextSelection.collapsed(offset: 15, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 13, end: 15))

</div><br /></b>
</details>

https://user-images.githubusercontent.com/2773578/236177474-4128bc6b-5244-4276-925e-ce1f0e39e5eb.mp4

2. If I write a new word, press space, then press backspace twice, then put the word's original last character, then press space and this word disappears. 

<details>
  <summary>Delete Word - Logs Fine</summary>
<meta charset="utf-8"><b style="font-weight:normal;" id="docs-internal-guid-e633bd63-7fff-3e85-e402-6f8f5d25d7e7"><div dir="ltr" style="margin-left:0pt;" align="left">

I  (52.179) textfield.android > FINE: User tapped down
I  (52.236) textfield.android > FINE: User released a tap
I  (52.242) textfield.android > INFO: Attaching TextInputClient to TextInput
I  (52.243) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy├, selection: TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (52.243) textfield.ime > FINE: Is attached to input client? true
I  (52.537) textfield.android > FINE: Intercepting single tap
I  (5.016) textfield.ime > FINE: Received text editing deltas from platform...
I  (5.016) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#23670(oldText: Good morning guy, selection: TextSelection.collapsed(offset: 16, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (5.017) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#399ab(oldText: Good morning guy, textInserted:, insertionOffset: 16, selection: TextSelection.collapsed(offset: 17, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (5.018) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy ├, selection: TextSelection.collapsed(offset: 17, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (6.277) textfield.ime > FINE: Received text editing deltas from platform...
I  (6.277) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#1a872(oldText: Good morning guy, textInserted: h, insertionOffset: 17, selection: TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 18))
I  (6.281) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy h├, selection: TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 18))
I  (6.813) textfield.ime > FINE: Received text editing deltas from platform...
I  (6.814) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#211c7(oldText: Good morning guy h, textInserted: e, insertionOffset: 18, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 19))
I  (6.815) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy he├, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 19))
I  (7.577) textfield.ime > FINE: Received text editing deltas from platform...
I  (7.578) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#578fa(oldText: Good morning guy he, textInserted: y, insertionOffset: 19, selection: TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 20))
I  (7.584) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy hey├, selection: TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 20))
I  (8.582) textfield.ime > FINE: Received text editing deltas from platform...
I  (8.583) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#fef5c(oldText: Good morning guy hey, textInserted:, insertionOffset: 20, selection: TextSelection.collapsed(offset: 21, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (8.584) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy hey ├, selection: TextSelection.collapsed(offset: 21, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (9.534) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy hey├, selection: TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (10.552) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy he├, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (11.494) textfield.ime > FINE: Received text editing deltas from platform...
I  (11.495) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#10887(oldText: Good morning guy he, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (11.495) textfield.ime > FINE: Processing selection/composing change: TextEditingDeltaNonTextUpdate#ab047(oldText: Good morning guy he, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 19))
I  (11.496) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy he├, selection: TextSelection.collapsed(offset: 19, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 19))
I  (11.498) textfield.ime > FINE: Received text editing deltas from platform...
I  (11.498) textfield.ime > FINE: Processing insertion: TextEditingDeltaInsertion#3311c(oldText: Good morning guy he, textInserted: y, insertionOffset: 19, selection: TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 20))
I  (11.499) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy hey├, selection: TextSelection.collapsed(offset: 20, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: 17, end: 20))
I  (11.500) textfield.ime > FINE: Received text editing deltas from platform...
I  (11.500) textfield.ime > FINE: Received text editing deltas from platform...
I  (12.340) textfield.ime > FINE: Processing replacement: TextEditingDeltaReplacement#b7103(oldText: Good morning guy hey, textReplaced: hey, replacementText:, replacedRange: TextRange(start: 17, end: 20), selection: TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
I  (12.341) textfield.ime > FINE: Sending TextEditingValue to platform: TextEditingValue(text: ┤Good morning guy  ├, selection: TextSelection.collapsed(offset: 18, affinity: TextAffinity.downstream, isDirectional: false), composing: TextRange(start: -1, end: -1))
 

</div><br /></b>
</details>

https://user-images.githubusercontent.com/2773578/236179038-e1bd4ccd-3ce5-4360-b000-f8202880fe42.mp4

### PR DESCRIPTION
**Pull request by angelosilvestre**
[SuperTextField][Android] Fix glitch with predictive text. Resolves #1113

On Android, some glitches are happening when "Predictive Text" is enabled on the Samsung keyboard. For example, in the SuperTextField Demo, after typing "you" at the end of the text, pressing backspace, typing "u" and pressing space, the word "you" is deleted.

I was able to reproduce this issue.

Currently, whenever we apply the list of deltas, we send our current text editing value to the OS and it seems that this is causing the issue. According to [this comment](https://github.com/flutter/flutter/issues/113253#issuecomment-1569238178), it seems we don't need to send the value to the OS if our editing value is the same as the OS thinks it is.

This PR changes `ImeAttributedTextEditingController` to store the IME current text editing value. Now, after applying the deltas, we check if the resulting value is the same as the OS thinks it is and we will only send the value to the OS if they are different. 

With this change I wasn't able to reproduce the issue anymore.

### CODE REVIEW

**Comment by angelosilvestre** [comment]
@ibrahimdevs Could you please try this PR and see if it solves the problem for you?

**Comment by angelosilvestre** [comment]
@matthew-carroll There is a failing test, but this test ensures that we send the editing value back to the OS when we receive a delta. If this PR is accepted we'll need to remove that test.

**Comment by matthew-carroll** [comment]
@angelosilvestre You're saying that exactly one test failed, and that test is checking for the exact behavior that you want to remove? If so, do we also have a test that verifies we send the delta back in the other situations, when we're supposed to?

**Comment by angelosilvestre** [comment]
@matthew-carroll The test only cares about the composing region. It tests that, given that we send back the value to the OS, we send it with the correct composing region.

It was added in https://github.com/superlistapp/super_editor/pull/983. Compound characters insertion still works correctly with the current PR.

**Comment by matthew-carroll** [comment]
It seems like that test was added for a reason. We shouldn't remove a test just because. We need to update the test so that it continues to confirm the part it cares about, while also complying with our new expectations.

**Comment by angelosilvestre** [comment]
@matthew-carroll With the changes from this PR, in most cases, we won't be sending the value back to the OS.

If the current text is "a" and the user types "b", after applying the delta we end up with the same value as the IME thinks it is ("ab").  Thus, we don't send it back. This will be true for all cases where we don't apply any special treatment after applying a delta (for example, prepending a placeholder, as we do in super_editor). 

I modified the test to use a custom controller which modifies the text when we insert a character to demonstrate this. This way, is possible to write a test that, given we need to send the value to the IME, we send it with the correct composing region.

File: super_editor/test/super_textfield/super_textfield_ime_test.dart [hunk 1]
```diff
@@ -73,22 +73,67 @@ void main() {
...
+          // Ensure we don't send the value back to the OS.
+          //
+          // As both us and the IME agree on what's the current editing value, we don't need to send it back.
+          expect(sentToPlatform, false);
+        });
+
+        testWidgetsOnAllPlatforms('and send editing value back to IME if it doesn\'t match the expected value',
+            (tester) async {
+          final controller = ImeAttributedTextEditingController(
+            controller: _ObscuringTextController(text: AttributedText(text: 'n')),
```
Comment (matthew-carroll) [inline super_editor/test/super_textfield/super_textfield_ime_test.dart:112]: Can you explain why you need to start with existing text for this test to work?

File: super_editor/test/super_textfield/super_textfield_ime_test.dart [hunk 2]
```diff
@@ -73,22 +73,67 @@ void main() {
...
+              if (methodCall.method == 'TextInput.setEditingState') {
+                sentToPlatform = true;
+                composingBase = methodCall.arguments["composingBase"];
+                composingExtent = methodCall.arguments["composingExtent"];
+              }
+              return null;
+            },
+          );
 
           // Simulate the user begining the input of the compound character 'ã'.
```
Comment (matthew-carroll) [inline super_editor/test/super_textfield/super_textfield_ime_test.dart:138]: This test seems to be re-using the compound character test from before, but it's not clear to me why they're related. If we want to test that we send updated values to the platform when they don't match, is there any reason that a compound character is required for that?

Also, the test that was turned into this one seemed to be focused on checking the composing region. I see that there are composing region assertions in this test, too, but are those assertions relevant to the goal of this test? Should we have a test dedicated to the composing region behavior, like before?

File: super_editor/test/super_textfield/super_textfield_ime_test.dart [hunk 3]
```diff
@@ -446,3 +499,64 @@ Widget _buildScaffold({
     ),
   );
 }
+
+/// An [AttributedTextEditingController] that uppon insertion replaces every character
+/// but the last one with a "*".
```
Comment (matthew-carroll) [inline super_editor/test/super_textfield/super_textfield_ime_test.dart:504]: Can you explain why we need this controller in this test suite?

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 1 above)
Comment (angelosilvestre) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:112]: There isn't a particular reason, we could also type the `n` instead of using as a starting text.

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 2 above)
Comment (angelosilvestre) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:138]: > is there any reason that a compound character is required for that?

The compound character isn't required for this test.

> Should we have a test dedicated to the composing region behavior, like before?

The problem is that we need to change the test in order to send the updated value to the IME. In the existing test, as the text doesn't change on our side, we don't send it back to the IME, so we can't check the composing region.

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 3 above)
Comment (angelosilvestre) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:504]: This controller is a way to change the text when the delta is processed, resulting in a different text from what the IME expects. This causes us to send the value back to the IME.

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 1 above)
Comment (matthew-carroll) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:112]: But what's the "n" for?

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 2 above)
Comment (matthew-carroll) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:138]: I understand that the existing test didn't change the content. But I'm confused as to why there isn't a minor adjustment we can make to that original test so that we do trigger an update, and once again confirm the change to the composing region?

Unless I'm missing something, it seems like there are 3 things to test, but we only have 2 tests. Also, at least one of those tests is based on details from the original test that aren't relevant to the new test. I want to make sure we're clearly testing the behaviors that we think we support, and I'd also like for each test to do the minimum work needed to confirm the relevant behavior.

File: super_editor/test/super_textfield/super_textfield_ime_test.dart (hunk 3 above)
Comment (matthew-carroll) [reply super_editor/test/super_textfield/super_textfield_ime_test.dart:504]: Can you add an appropriate explanation to the controller so that future readers know why it's there?

**Review by matthew-carroll — APPROVED** [review]
LGTM
